src/plugins/hidsim/hidsim.cpp

```diff
@@ -0,0 +1,160 @@
+ *                                                                         *
+ * Source code also allows you to port DRAKVUF to new platforms, fix bugs, *
+ * and add new features.  You are highly encouraged to submit your changes *
+ * on https://github.com/tklengyel/drakvuf, or by other methods.           *
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Jan Gruber.                                    *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+#include <libdrakvuf/libdrakvuf.h>  /* eprint_current_time */
+#include "../private.h"             /* PRINT_DEBUG */
+#include "hidsim.h"
+
+/* Infers socket path from drakvuf's actual domID */
+std::string construct_sock_path(drakvuf_t drakvuf)
+{
+    /* Retrieves domid as string */
+    std::string sock_path(SOCK_STUB);
+    sock_path.append(std::to_string(drakvuf_get_dom_id(drakvuf)));
+    return sock_path;
+}
+
+hidsim::hidsim(drakvuf_t drakvuf, const hidsim_config* config)
+{
+
+    if (config->template_fp)
+    {
+        this->template_path = config->template_fp;
+        PRINT_DEBUG("[HIDSIM] Using template file: %s\n", this->template_path.c_str());
+    }
+
+    /* Constructs path to Unix domain socket of Xen guest under investigation */
+    this->sock_path = construct_sock_path(drakvuf);
+    PRINT_DEBUG("[HIDSIM] Using Unix domain socket: %s\n", this->sock_path.c_str());
+    this->has_to_stop = false;
+
+    /* Starts worker thread */
+    this->t = std::thread(hid_inject, sock_path.c_str(), template_path.c_str(), &has_to_stop);
+    PRINT_DEBUG("[HIDSIM] HID injection started\n");
+}
+
+hidsim::~hidsim()
+{
+    if (!this->m_is_stopping)
```

Comment:
Why do we guard here against calling destructor multiple times? I don't think this is going to happen unless there is some corruption.

Reply:
The thought behind it, was that it's theoretically possible to call it multiple times, but you are right, should never happen unless corruption. I removed it to simplify the code.